src/app/reporting/SynchronizedReportSchedulerImpl.cpp

```diff
@@ -0,0 +1,178 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/InteractionModelEngine.h>
+#include <app/reporting/SynchronizedReportSchedulerImpl.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+namespace chip {
+namespace app {
+namespace reporting {
+
+using namespace System::Clock;
+using ReadHandlerNode = ReportScheduler::ReadHandlerNode;
+
+CHIP_ERROR SynchronizedReportSchedulerImpl::ScheduleReport(Timeout timeout, ReadHandlerNode * node)
+{
+    // Cancel Report if it is currently scheduled
+    mTimerDelegate->CancelTimer(this);
+    ReturnErrorOnFailure(mTimerDelegate->StartTimer(this, timeout));
+    mTestNextReportTimestamp = mTimerDelegate->GetCurrentMonotonicTimestamp() + timeout;
+
+    return CHIP_NO_ERROR;
+}
+
+void SynchronizedReportSchedulerImpl::CancelReport(ReadHandler * aReadHandler)
+{
+    // We don't need to take action on the handler, since the timer is common here
+    mTimerDelegate->CancelTimer(this);
+}
+
+void SynchronizedReportSchedulerImpl::UnregisterReadHandler(ReadHandler * aReadHandler)
+{
+    // Verify list is populated and handler is not null
+    VerifyOrReturn((!mReadHandlerList.Empty() || (nullptr == aReadHandler)));
+
+    ReadHandlerNode * removeNode = FindReadHandlerNode(aReadHandler);
+    // Nothing to remove if the handler is not found in the list
+    VerifyOrReturn(nullptr != removeNode);
+
+    mReadHandlerList.Remove(removeNode);
+    mNodesPool.ReleaseObject(removeNode);
+
+    if (mReadHandlerList.Empty())
+    {
+        // Only cancel the timer if there are no more handlers registered
+        CancelReport(aReadHandler);
+    }
+}
+
+/// @brief Checks if the timer is active for the given ReadHandler. Since all read handlers are scheduled on the same timer, we
+/// check if the node is in the list and if the timer is active for the ReportScheduler
+bool SynchronizedReportSchedulerImpl::IsReportScheduled(ReadHandler * aReadHandler)
+{
+    ReadHandlerNode * node = FindReadHandlerNode(aReadHandler);
+    VerifyOrReturnValue(nullptr != node, false);
+    return mTimerDelegate->IsTimerActive(this);
+}
+
+/// @brief Find the smallest maximum interval possible and set it as the common maximum
+/// @return NO_ERROR if the smallest maximum interval was found, error otherwise, INVALID LIST LENGTH if the list is empty
+CHIP_ERROR SynchronizedReportSchedulerImpl::FindNextMaxInterval()
+{
+    VerifyOrReturnError(!mReadHandlerList.Empty(), CHIP_ERROR_INVALID_LIST_LENGTH);
+    System::Clock::Timestamp now      = mTimerDelegate->GetCurrentMonotonicTimestamp();
+    System::Clock::Timestamp earliest = now + Seconds16(kSubscriptionMaxIntervalPublisherLimit);
+
+    for (auto & iter : mReadHandlerList)
+    {
+        if (iter.GetMaxTimestamp() < earliest && iter.GetMaxTimestamp() > now)
```

Comment:
The `> now` is to prevent setting a nextMax in the past, maybe `>= now` would be better? A max-interval close to expire would be picked as our next max no?

Reply:
> The > now is to prevent setting a nextMax in the past

I understand what it's doing, but what does that situation really mean, when we have a max timestamp in the past?  Doesn't it mean that we have already missed out deadline and should report ASAP and hope the MRP resend slack covers up us missing the deadline?  Again, followup is fine for this; this needs some really careful analysis.

Reply:
We will have a max in the past when we are currently through an engine run with multiple available handlers.

Say both readhandler1 and readhandler2 are reportable on timeout. 
We start a run.
Readhandler1's report is sent, we update its intervals and we the Calculate the next timeout.
ReadHandler2's max is in the past, and will be the earliest, so we ignore it.

Issue created here : https://github.com/project-chip/connectedhomeip/issues/27996